### ISSUE TITLE
Introduce a variable for lhs of assignment

### ISSUE THREAD

**Issue by xiaq** [elves/elvish#242]
Elvish does not have self-operators like `+=`. However, this can be compensated by having a special readonly variable that refers to the current value of the lhs of the assignment. For instance:

```
very-long-variable-name=(+ 10 $lhs)
```

Although only two characters shorter, it also makes it easier to manipulate `$paths`:

```
paths=[~/bin $@lhs /usr/sbin]
```

**Comment by krader1961** [comment]
I do not think the added complexity of this proposal, both in terms of the implementation and documentation, justifies shortening statements used in relatively rare situations. At least not as proposed in the opening comment of this issue. It introduces an ad-hoc rule and means an elvish user cannot create a var named `lhs`.  If someone wants to augment a var such as `$paths` it is not an undue burden for them to use the var name in the expression on the RHS of the assignment.

**Comment by hanche** [comment]
I think a better argument in favour of the proposal is when the left hand side has a form like `foo[(blah)]`, where the `blah` function may involve a costly computation, possibly with side effects. But of course, even this is fairly easily handled right now, so it's still not a really strong argument:

`i=(blah) foo[$i] = (some-computation-involving $foo[$i])`

**Comment by zzamboni** [comment]
After reading the comments, I agree with @krader1961 that this probably introduces more complexity than it's worth. Ultimately any shortcut just for the sake of shortness introduces "magic" that makes it harder for someone to understand the code.

**Comment by krader1961** [comment]
FWIW, There are people in the fish shell community who take "DRY" (don't repeat yourself) to extreme lengths. One example is people felt quite strongly that having to type `set path $path /new/dir` was unacceptable since you have to specify the var name twice. I eventually threw up my arms in defeat and implemented `set -a path /new/dir`, and corresponding `set -p` for prefixing. But that didn't need to introduce any magic and the is mostly self-explanatory.

**Comment by xiaq** [comment]
The motivation for this issue is mainly to make it easier to mutate deeply nested; consider `x[deeply][nested][element] = (+ $x[deeply][nested][element] 1)`. In languages with augmented assignments, this can be written as `x[deeply][nested][element] += 1`, but that syntax presumes the existence of operators (which Elvish does not have; `+` is just a command). Hence the idea of getting something on par with augmented assignments by introducing a `$lhs`.

That said, I am not sure this is a good idea, the "magicness" is indeed a bit distasteful. Most Lisps manage fine without augmented assignments or "LHS variable", so Elvish can perhaps also live without it.

**Comment by xiaq** [comment]
Another design idea is to support a generalized version of augmented assignment, making `varname <cmd>= args...` a shorthand for `varname = (<cmd> varname args...)`, but this makes every word ending with `=` special, and conflicts with a few things:

* When supplying an empty value for a variable to `make`, such as `make ENV= CC=clang` - which will be interpreted as `make = (ENV $make CC=clang)`;

* When supplying an empty value for options, such as `cmd -o= arg`, which will now be interpreted as `cmd = (-o $cmd arg)`.

So this isn't a good idea either, at least not in this form.

**Comment by krader1961** [comment]
Okay, the deeply nested element example is a bit more compelling. It also has some interesting wrinkles. Consider `x[deeply][nested][element] = [lhs]{ put $lhs }`. Legal? Apparently yes given the rules discussed so far. But it probably shouldn't be assuming the replacement is not done in the lambda argument list but is done in the body. Or, perhaps lambdas represent a special-case that inhibits the LHS mechanism. Now you've got a special-case that has to be documented and will cause confusion.

Counter-proposal: What about making it possible on the RHS to indicate that a var being used should also be updated with the output of the command? Something like `+ 17 $=x[deeply][nested][element]`. I'm not a fan of this approach because it tends to make people want to add more and more magic modifiers of a var dereference. C.f., bash, ksh, zsh. On the other hand Xiaq has good design sensibilities so that seems unlikely to happen.

**Comment by SolitudeSF** [comment]
some basic metaprogramming features would be nice for this.

**Comment by hanche** [comment]
A likely misunderstanding of the latest comment by @krader1961 made me think. What if we introduced a new assignment syntax with new semantics? For the syntax, an infix operator like `:=` or perhaps even `<=` to strongly signal that something unusual is going on, and for the semantics, on the right you put a lambda which is fed the old value and then produces the new value. For example to increment a value:

`x[deeply][nested][element] <= [old]{ + $old 1 }`

**Comment by xiaq** [comment]
@hanche I think what you proposed is the most sensible version so far, if slightly verbose. Moreover, a special `mutate` command would look cleaner:

```
mutate x[deeply][nested][element] [old]{ + $old 1 }
```

In fact, such a command can be more than syntactical sugar, it can function as an **atomic** mutation command and solve the problem of #888.

**Comment by xiaq** [comment]
Closing this one as the original use case is now addressed in #1003.
